lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -4245,6 +4245,96 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
     return BM->addInstTemplate(internal::OpMaskedScatterINTEL, Ops, BB,
                                nullptr);
   }
+  case Intrinsic::is_fpclass: {
+    // There is no direct counterpart for the intrinsic in SPIR-V, hence
+    // we need to emulate it's work by sequence of other instructions
+    SPIRVType *Ty = transType(II->getType());
+    auto *InputFloat = transValue(II->getArgOperand(0), BB);
+    uint64_t FPClass =
+        cast<ConstantInt>(II->getArgOperand(1))->getZExtValue();
+    std::vector<SPIRVValue *> ResultVec;
+    std::vector<SPIRVId> Ops = {InputFloat->getId()};
+    // Adds test for Negative/Positive values
+    auto GetNegPosInstTest = [&](SPIRVValue *TestInst, bool IsNegative) {
+      auto *SignBitTest =
+          BM->addInstTemplate(OpSignBitSet, {TestInst->getId()}, BB, Ty);
+      if (IsNegative)
+        return SignBitTest;
+      return BM->addInstTemplate(OpLogicalNot, {SignBitTest->getId()}, BB, Ty);
+    };
+    // Integer paramter of the intrinsic is combined from several bit masks
+    // referenced in FPClassTest enum from FloatingPointMode.h in LLVM.
+    // Since a single intrinsic can provide multiple tests - here we might end
+    // up adding several sequences of SPIR-V instructions
+    if (FPClass & fcSNan || FPClass & fcQNan) {
+      // There are no signaling and quite NaN instructions in SPIR-V, so map
+      // both of the tests on OpIsNan
+      ResultVec.emplace_back(BM->addInstTemplate(OpIsNan, Ops, BB, Ty));
+    }
+    if (FPClass & fcNegInf || FPClass & fcPosInf) {
+      auto *TestIsInf = BM->addInstTemplate(OpIsInf, Ops, BB, Ty);
+      if (FPClass & fcNegInf && FPClass & fcPosInf)
+        // Map on OpIsInf if we have both Inf test bits set
+        ResultVec.emplace_back(TestIsInf);
+      else
+        // Map on OpIsInf with following check for sign bit
+        ResultVec.emplace_back(GetNegPosInstTest(TestIsInf,
+                                                 FPClass & fcNegInf));
+    }
+    if (FPClass & fcNegNormal || FPClass & fcPosNormal) {
+      auto *TestIsNormal = BM->addInstTemplate(OpIsNormal, Ops, BB, Ty);
+      if (FPClass & fcNegNormal && FPClass & fcPosNormal)
+        // Map on OpIsNormal if we have both Normal test bits set
+        ResultVec.emplace_back(TestIsNormal);
+      else
+        // Map on OpIsNormal with following check for sign bit
+        ResultVec.emplace_back(GetNegPosInstTest(TestIsNormal,
+                                                 FPClass & fcNegNormal));
+    }
+    if (FPClass & fcNegSubnormal || FPClass & fcPosSubnormal) {
+      // If value is both not NaN and Normal, then it's subnormal
+      // TODO: it's quite silly check, can we figure out something better?
```

Comment:
@svenvh @asudarsa @bashbaug may I ask you to take a look? I assumed, that subnormal test should be test for not Nan and not Normal, though may be you see better ways to perform that.

Reply:
Simplest check here is to check if the exponent bits are minimum exponent, and mantissa is nonzero. Looking at the TargetLowering variant, this can be reduced further to `unsigned(mask_sign(V) - 1) < all_ones_mantissa`